core/shared/src/main/scala/zio/ZIOMetric.scala

```diff
@@ -582,14 +706,40 @@ object ZIOMetric {
      * Increments the counter associated with the specified value by one.
      */
     def observe(value: String)(implicit trace: ZTraceElement): UIO[Any] =
-      setCount.observe(value)
+      withSetCount(_.observe(value))
 
     /**
-     * Returns the number of occurences of every value observed by this set
+     * Returns the number of occurrences of every value observed by this set
      * count.
      */
     def occurrences(implicit trace: ZTraceElement): UIO[Chunk[(String, Long)]] =
-      setCount.occurrences
+      withSetCount(_.occurrences)
+
+    /**
+     * Converts this set count metric to one where the tags depend on the
+     * measured effect's result value
+     */
+    def taggedWith(f: A => Chunk[MetricLabel]): ZIOMetric[A] = {
+      if (self.setCountRef eq null) {
+        self.setCountRef = ZFiberRef.unsafeMake(self.setCount)
+        self.setCount = null
+      }
+      new ZIOMetric[A] {
+        override def apply[R, E, A1 <: A](zio: ZIO[R, E, A1])(implicit trace: ZTraceElement): ZIO[R, E, A1] =
+          self.apply(zio.tap(changeSetCount))
+
+        private def changeSetCount(value: A)(implicit trace: ZTraceElement): UIO[Unit] =
+          self.setCountRef.update { setCount =>
```

Comment:
You can improve performance by factoring this lambda out into a constant lambda inside the class. That way, it won't be created each time `changeSetCount` is invoked.